linenoise/linenoise/linenoise.h

```diff
@@ -0,0 +1,335 @@
+
+/**
+ * @brief Sets whether an empty line is allowed to be returned.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param empty_line true to allow empty lines, false to disallow.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_set_empty_line(linenoise_handle_t handle, bool empty_line);
+
+/**
+ * @brief Checks whether the instance allows returning an empty line.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param is_empty_line Pointer to bool that will be set with result.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_is_empty_line(linenoise_handle_t handle, bool *is_empty_line);
+
+/**
+ * @brief Enables or disables multi-line editing.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param multi_line true to enable multi-line input, false to disable.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_set_multi_line(linenoise_handle_t handle, bool multi_line);
+
+/**
+ * @brief Checks if multi-line editing is enabled.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param is_multi_line Pointer to bool that will be set with result.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_is_multi_line(linenoise_handle_t handle, bool *is_multi_line);
+
+/**
+ * @brief Enables or disables dumb mode (disables line editing features).
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param dumb_mode true to enable dumb mode, false to disable.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_set_dumb_mode(linenoise_handle_t handle, bool dumb_mode);
+
+/**
+ * @brief Checks if dumb mode is enabled.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param is_dumb_mode Pointer to bool that will be set with result.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_is_dumb_mode(linenoise_handle_t handle, bool *is_dumb_mode);
+
+/**
+ * @brief Sets the maximum length of the command line buffer.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param length Maximum length in bytes.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_set_max_cmd_line_length(linenoise_handle_t handle, size_t length);
+
+/**
+ * @brief Gets the maximum allowed command line buffer length.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param max_cmd_line_length Pointer to receive the max length.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_get_max_cmd_line_length(linenoise_handle_t handle, size_t *max_cmd_line_length);
+
+/*------------------------------------------------------------------------------------------------
+ *OLD LINENOISE API
+ *------------------------------------------------------------------------------------------------*/
+
+typedef struct linenoise_completions linenoiseCompletions;
```

Comment:
I'd recommend having the compatibility layer in a different header file from the "new" API.

Components which use the old API, can include, for example, "linenoise/linenoise.h" (as they used to).

Components which use the new API can include a different file, such as "esp_linenoise.h", and only get the new functions, without the old functions polluting the global namespace.

Reply:
topology updated. I added the new linenoise to esp_linenoise.c/h files